### ISSUE TITLE
SwiftIfConfig: Create a "fixed" or "static" build configuration that is Codable

### ISSUE THREAD

**Issue by DougGregor** [swiftlang/swift-syntax#2884]
### Description

The SwiftIfConfig library has a `BuildConfiguration` protocol, but there is no concrete implementation of the `BuildConfiguration` protocol in the library. The only `BuildConfiguration` in normal use right now is buried in the Swift compiler itself, to query the configuration of the compiler.

We should add some kind of "fixed" or "static" build configuration that conforms to `BuildConfiguration` and is `Codable` so it can easily be saved and restored. Ideas for such a configuration:

* Use it for some of the unit tests instead of the bare-bones testing build configuration we have now
* Extend the compiler and/or SourceKit with an option to dump out JSON in a format that's compatible with this type, so tools can query the build configuration once and then apply it in other contexts such as an IDE (e.g., in SourceKit-LSP)
* Extend swift-parser-cli with an option to apply a build configuration file (the serialized JSON), removing all of the inactive `#if` regions based on that configuration (SwiftIfConfig already has an API for this)

**Comment by ahoppen** [comment]
Synced to Apple’s issue tracker as rdar://138352412

**Comment by broken-circle** [comment]
Took a look at this and noticed the work is mostly done. `StaticBuildConfiguration` already exists, is `Codable`, and is in use across the test suite. `swift-parser-cli` has a `--build-configuration` option loading a JSON-encoded `StaticBuildConfiguration` and applying `removingInactive(in:)` to the parsed tree. 

I opened #3333 to fix some small typos I noticed in the public documentation.

The only remaining work is extending the compiler/SourceKit to emit the JSON. Since that's outside this repo, this issue can probably be closed now.

**Comment by DougGregor** [comment]
I agree that can be closed. I'd forgotten I opened it when adding `StaticBuildConfiguration`.
